packages/iov-bns/src/types.ts

```diff
@@ -176,3 +194,153 @@ export const ensure = <T>(maybe: T | null | undefined, msg?: string): T => {
   }
   return maybe;
 };
+
+// transactions
+
+/**
+ * BNS transaction kinds.
+ *
+ * Raw values are used for type detection only and can change at any time.
+ */
+export enum TransactionKind {
+  AddAddressToUsername = "add_address_to_username",
+  Send = "send",
+  /** @deprecated see SetNameTx */
+  SetName = "set_name",
+  SwapOffer = "swap_offer",
+  SwapCounter = "swap_counter",
+  SwapClaim = "swap_claim",
+  SwapTimeout = "swap_timeout",
+  RegisterBlockchain = "register_blockchain",
+  RegisterUsername = "register_username",
+  RemoveAddressFromUsername = "remove_address_from_username",
+}
+
+const bnsDomain = "bns";
+
+export interface BnsTx extends UnsignedTransaction {
```

Comment:
No union type?
I found those union types (along with ifs/switches) to be very nice to check all cases.
Here it seems a bit more manual boilerplate with the type detection functions everywhere, which may have bugs (rather than `tx.kind == ???`). Wonder why you made this choice.

Reply:
It was not explicit choice other than: let's get rid of centralized lists. But inside a blockchain package we can of cource have those lists.

We need the manual boilerplate type detection functions because they check two things:

1. is BnsTx
2. is the specific kind of BnsTx we are looking for

and we don't want to have those two steps in the caller code.

So both ways work. I brought back the union type in https://github.com/iov-one/iov-core/pull/605/commits/fd0dfd28787014712cdb102f1441759715ecbfab